src/audio/smart_amp_maxim.c

```diff
@@ -0,0 +1,719 @@
+{
+	struct sof_dsm_struct_t *hsof_dsm;
+	int mem_sz[3];
+	int size;
+
+	if (sad->sof_dsm_handle)
+		return 0;
+
+	/* sof dsm handle */
+	mem_sz[0] = sizeof(struct sof_dsm_struct_t);
+	sad->sof_dsm_handle = smart_amp_malloc(mem_sz[0]);
+	if (!sad->sof_dsm_handle)
+		goto err;
+
+	hsof_dsm = sad->sof_dsm_handle;
+
+	/* buffer : sof -> sof dsm ff */
+	size = SOF_FF_BUF_DB_SZ * sizeof(int32_t);
+	hsof_dsm->buf.sof_a_frame_in = smart_amp_malloc(size);
+	if (!hsof_dsm->buf.sof_a_frame_in)
+		goto err;
+	mem_sz[2] = size;
+
+	/* buffer : sof <- sof dsm ff */
+	size = SOF_FF_BUF_DB_SZ * sizeof(int32_t);
+	hsof_dsm->buf.sof_a_frame_out = smart_amp_malloc(size);
+	if (!hsof_dsm->buf.sof_a_frame_out)
+		goto err;
+	mem_sz[2] += size;
+
+	/* buffer : sof -> sof dsm fb */
+	size = SOF_FB_BUF_DB_SZ * sizeof(int32_t);
+	hsof_dsm->buf.sof_a_frame_iv = smart_amp_malloc(size);
+	if (!hsof_dsm->buf.sof_a_frame_iv)
+		goto err;
+	mem_sz[2] += size;
+
+	/* buffer : sof dsm ff stage */
+	size = DSM_FF_BUF_DB_SZ * sizeof(int32_t);
+	hsof_dsm->buf.stage = smart_amp_malloc(size);
+	if (!hsof_dsm->buf.stage)
+		goto err;
+	mem_sz[2] += size;
+
+	/* buffer : sof dsm fb stage */
+	size = DSM_FB_BUF_DB_SZ * sizeof(int32_t);
+	hsof_dsm->buf.stage_fb = smart_amp_malloc(size);
+	if (!hsof_dsm->buf.stage_fb)
+		goto err;
+	mem_sz[2] += size;
+
+	/* buffer : sof dsm -> dsm ff in */
+	size = DSM_FF_BUF_SZ * sizeof(int16_t);
+	hsof_dsm->buf.input = smart_amp_malloc(size);
+	if (!hsof_dsm->buf.input)
+		goto err;
+	mem_sz[2] += size;
+
+	/* buffer : sof dsm <- dsm ff out */
+	size = DSM_FF_BUF_SZ * sizeof(int16_t);
+	hsof_dsm->buf.output = smart_amp_malloc(size);
+	if (!hsof_dsm->buf.output)
+		goto err;
+	mem_sz[2] += size;
+
+	/* buffer : sof dsm -> dsm fb voltage */
+	size = DSM_FF_BUF_SZ * sizeof(int16_t);
+	hsof_dsm->buf.voltage = smart_amp_malloc(size);
+	if (!hsof_dsm->buf.voltage)
+		goto err;
+	mem_sz[2] += size;
+
+	/* buffer : sof dsm -> dsm fb current */
+	size = DSM_FF_BUF_SZ * sizeof(int16_t);
+	hsof_dsm->buf.current = smart_amp_malloc(size);
+	if (!hsof_dsm->buf.current)
+		goto err;
+	mem_sz[2] += size;
+
+	/* buffer : sof dsm ff in : variable length -> fixed length */
+	size = DSM_FF_BUF_DB_SZ * sizeof(int32_t);
+	hsof_dsm->buf.ff.buf = smart_amp_malloc(size);
+	if (!hsof_dsm->buf.ff.buf)
+		goto err;
+	mem_sz[2] += size;
+
+	/* buffer : sof dsm ff out : variable length <- fixed length */
+	size = DSM_FF_BUF_DB_SZ * sizeof(int32_t);
+	hsof_dsm->buf.ff_out.buf = smart_amp_malloc(size);
+	if (!hsof_dsm->buf.ff_out.buf)
+		goto err;
+	mem_sz[2] += size;
+
+	/* buffer : sof dsm fb : variable length -> fixed length */
+	size = DSM_FB_BUF_DB_SZ * sizeof(int32_t);
+	hsof_dsm->buf.fb.buf = smart_amp_malloc(size);
+	if (!hsof_dsm->buf.fb.buf)
+		goto err;
+	mem_sz[2] += size;
+
+	comp_dbg(dev, "[DSM] sof:%p (%d bytes), dsm:%p (%d bytes)",
+		 (uintptr_t)hsof_dsm, mem_sz[0] + mem_sz[2],
+		 (uintptr_t)hsof_dsm->dsmhandle, mem_sz[1]);
+
+	return 0;
+err:
+	return -1;
```

Comment:
-ENOMEM 
Please also report this error via trace error call.

Reply:
Agreed. Shall fix this.